lib/codecs/src/encoding/format/cef.rs

```diff
@@ -0,0 +1,483 @@
+    pub cef: CefSerializerOptions,
+}
+
+impl CefSerializerConfig {
+    /// Creates a new `CefSerializerConfig`.
+    pub const fn new(cef: CefSerializerOptions) -> Self {
+        Self { cef }
+    }
+
+    /// Build the `CefSerializer` from this configuration.
+    pub fn build(&self) -> Result<CefSerializer, BuildError> {
+        let device_vendor = if let Some(device_vendor) = self.cef.device_vendor.clone() {
+            escape_header(device_vendor)
+        } else {
+            String::from("Datadog")
+        };
+        if device_vendor.len() > 63 {
+            return Err(format!(
+                "device_vendor exceed 63 characters limit: actual {}",
+                device_vendor.len()
+            )
+            .into());
+        };
+
+        let device_product = if let Some(device_product) = self.cef.device_product.clone() {
+            escape_header(device_product)
+        } else {
+            String::from("Vector")
+        };
+        if device_product.len() > 63 {
+            return Err(format!(
+                "device_product exceed 63 characters limit: actual {}",
+                device_product.len()
+            )
+            .into());
+        };
+
+        let device_version = if let Some(device_version) = self.cef.device_version.clone() {
+            device_version
+        } else {
+            String::from("0") // Major version. TODO(nabokihms): find a way to get the actual vector version.
+        };
+        if device_version.len() > 31 {
+            return Err(format!(
+                "device_version exceed 31 characters limit: actual {}",
+                device_version.len()
+            )
+            .into());
+        };
+
+        let device_event_class_id =
+            if let Some(device_event_class_id) = self.cef.device_event_class_id.clone() {
+                escape_header(device_event_class_id)
+            } else {
+                String::from("Telemetry Event")
+            };
+        if device_event_class_id.len() > 1023 {
+            return Err(format!(
+                "device_event_class_id exceed 1023 characters limit: actual {}",
+                device_event_class_id.len()
+            )
+            .into());
+        };
+
+        for key in self.cef.extensions.keys() {
+            if !key.chars().all(|c| c.is_ascii_alphabetic()) {
+                // TODO (nabokihms): Output all invalid keys
+                return Err(format!("extension keys can only contain ascii alphabetical characters: invalid key '{}'", key).into());
+            }
+        }
+
+        let device = DeviceSettings::new(
+            device_vendor,
+            device_product,
+            device_version,
+            device_event_class_id,
+        );
+
+        Ok(CefSerializer::new(
+            self.cef.version.clone(),
+            device,
+            self.cef.severity.clone(),
+            self.cef.name.clone(),
+            self.cef.extensions.clone(),
+        ))
```

Comment:
It looks like the only thing really being built here is the device settings 🤔 Just seems wasteful to have all of these clones...

Reply:
Skipped here because I didn't get how to change it properly. I don't want to use time annotations on the Serializers struct, and it seems like cloning a small string once is not a big deal. 

If I'm wrong here, please guide me how to fix it properly.

Reply:
We can leave this is as is.